messaginginapp/src/main/java/io/customer/messaginginapp/state/InAppMessagingMiddlewares.kt

```diff
@@ -149,13 +149,46 @@ internal fun processMessages() = middleware<InAppMessagingState> { store, next,
         val notShownMessages = action.messages
             .filter { message ->
                 // filter out the messages that are already shown
-                // and the messages that have an elementId because we are not handling embedded messages
-                message.queueId != null && !store.state.shownMessageQueueIds.contains(message.queueId) && message.gistProperties.elementId == null
+                message.queueId != null && !store.state.shownMessageQueueIds.contains(message.queueId)
             }
             .distinctBy(Message::queueId)
             .sortedWith(compareBy(nullsLast()) { it.priority })
 
-        val messageToBeShownWithProperties = notShownMessages.firstOrNull { message ->
+        val modalMessages = notShownMessages.filter { it.gistProperties.elementId == null }
+        val inlineMessages = notShownMessages.filter { it.gistProperties.elementId != null }
+
+        val isCurrentMessageDisplaying = store.state.modalMessageState is ModalMessageState.Displayed
+        val isCurrentMessageBeingProcessed = store.state.modalMessageState is ModalMessageState.Loading
+
+        // update the state with the messages in the queue that are not shown
+        // because in the next steps we will check if there is a message to be shown and display them
+        next(InAppMessagingAction.ProcessMessageQueue(notShownMessages))
+
+        // Handle embedded messages
+        val inLineMessagesToBeShown = inlineMessages
+            .filter { message ->
+                val routeRule = message.gistProperties.routeRule
+                val currentRoute = store.state.currentRoute
+                // Check if the message matches the current route
+                when {
+                    routeRule == null -> true
+                    currentRoute == null -> false
+                    else -> runCatching { routeRule.toRegex().matches(currentRoute) }.getOrNull() ?: true
+                }
```

Comment:
Would be nice to move this to extension so route matching is done by single function for both modal and inline messages.